[DI - Step 1] 이지 미션 제출합니다.

skygl opened the pull request:
피드백 잘부탁드립니다~

skygl committed:
feat: BeanFactory.initialize() 구현

- 테스트 통과

skygl committed:
refactor: BeanFactory 리팩터링

- BeanFactoryUtils 리팩토링
 - getInjectedConstructor 변경
- 메서드 분리

skygl committed:
refactor: mvc 패키지에 di 적용

- UserAcceptanceTest Disabled 어노테이션을 사용하여 실행 대상 제외
- UserDao 싱글톤 해제

skygl committed:
feat: Stack을 이용하여 Bean 객체 생성시 순환 참조 체크

skygl committed:
refactor: nextstep.di.factory.error 패키지 생성

- 에러가 발생하는 빈들을 패키지에 모음
- 에러 케이스별 어노테이션 사용

skygl committed:
refactor: BeanScanner 생성자 파라미터 변경

- 필요 어노테이션을 인자로 전달받게 변경

kang-hyungu commented on nextstep-di/src/test/java/nextstep/di/factory/BeanFactoryErrorTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Test
+    void 기본_생성자만_있는_클래스_빈_생성_확인() {
+        Set<Class<?>> types = reflections.getTypesAnnotatedWith(GoodBean.class);
```
빈 팩토리 에러 발생 테스트를 위해 테스트용 어노테이션(GoodBean, BadBean, Circular)을 새로 만들 필요가 있었을까요?
테스트 코드를 샘플처럼 참고하는 경우도 있는데 실제 프레임워크에서 사용하는 어노테이션으로 예제를 보여주는게 좋지 않을까요?

skygl replied:
패키지 구조로 나누어서 테스트 하도록 수정했습니다 👍

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -29,8 +33,12 @@ public AnnotationHandlerMapping(Object... basePackage) {
     public void initialize() {
-        ControllerScanner controllerScanner = new ControllerScanner(basePackage);
-        Map<Class<?>, Object> controllers = controllerScanner.getControllers();
+        BeanScanner beanScanner = new BeanScanner(basePackage, Controller.class, Service.class, Repository.class);
```
빈 스캐너가 어떤 어노테이션을 탐색할지 어노테이션 핸들러 매핑이 정해주는건 적절해보이지 않네요
빈 스캐너가 어느 어노테이션을 찾아서 반환할지 알고 있어야하지 않을까요?

skygl replied:
빈 스캐너가 어노테이션을 갖도록 수정했습니다.

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -29,8 +33,12 @@ public AnnotationHandlerMapping(Object... basePackage) {
-        ControllerScanner controllerScanner = new ControllerScanner(basePackage);
-        Map<Class<?>, Object> controllers = controllerScanner.getControllers();
+        BeanScanner beanScanner = new BeanScanner(basePackage, Controller.class, Service.class, Repository.class);
+        BeanFactory beanFactory = beanScanner.getBeanFactory();
```
빈 팩토리가 빈 스캐너에 의존하는 관계가 됐네요
빈 스캐너가 Class를 반환해서 빈 팩토리에서 처리하도록 바꿔보는건 어떨까요?

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -29,8 +33,12 @@ public AnnotationHandlerMapping(Object... basePackage) {
+        BeanFactory beanFactory = beanScanner.getBeanFactory();
+        beanFactory.initialize();
+
+        Map<Class<?>, Object> controllers = beanFactory.getBeansWithAnnotation(Controller.class);
```
빈으로 등록된 클래스 목록만 받아와도 되지 않을까요?
생성된 객체는 빈팩토리가 알고 있으니 빈 팩토리에게 물어봐서 필요한 객체를 반환해도 되지 않을까요?

kang-hyungu commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -29,8 +33,12 @@ public AnnotationHandlerMapping(Object... basePackage) {
+        Map<Class<?>, Object> controllers = beanFactory.getBeansWithAnnotation(Controller.class);
+
         Set<Method> methods = getRequestMappingMethods(controllers.keySet());
         for (Method method : methods) {
```
for문을 메소드 추출하면 코드가 좀더 깔끔해질 것 같아요

kang-hyungu commented on slipp/src/test/java/slipp/controller/UserAcceptanceTest.java:
```diff
@@ -1,19 +1,21 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+@Disabled
```
인수 테스트에 문제가 있어서 disabled 처리한걸까요?

skygl replied:
서버를 켜줘야 테스트가 정상적으로 작동해서 Disabled 어노테이션을 붙였는데, 어노테이션을 제거하고 서버를 작동시키는 코드를 추가했습니다.

kang-hyungu commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -14,6 +20,8 @@
 
     private Map<Class<?>, Object> beans = Maps.newHashMap();
 
+    private CircularReferenceDetector circularReferenceDetector = new CircularReferenceDetector();
```
static 변수가 아니니 initialize 메서드가 실행될 때 객체 생성해도 되지 않을까요?

skygl replied:
'해당 변수가 initialize 메서드에서만 사용해서 initialize 메서드가 실행될 때 초기화를 해도 되지 않을까' 라는 의미가 맞을까요??

kang-hyungu replied:
네 static 변숫가 아닌데 미리 객체를 생성해둔 이유가 궁금했어요

kang-hyungu commented on nextstep-di/src/test/java/nextstep/di/factory/BeanFactoryTest.java:
```diff
@@ -53,4 +54,6 @@ public void di() throws Exception {
         log.debug("Scan Beans Type : {}", beans);
         return beans;
     }
+
```
빈 팩토리가 의도한대로 빈 반환 및 생성해주고 있는지 아래 테스트를 추가해서 확인해보아요!
```java
    @Test
    void UserRepository가_싱글_인스턴스가_맞는지_테스트() {
        final UserRepository beanFactoryUserRepository = beanFactory.getBean(JdbcUserRepository.class);

        final MyQnaService qnaService = beanFactory.getBean(MyQnaService.class);
        final UserRepository qnaServiceUserRepository = qnaService.getUserRepository();

        assertThat(beanFactoryUserRepository).isEqualTo(qnaServiceUserRepository);
    }
```

skygl replied:
인터페이스일 경우 인터페이스 클래스에 대해서 빈을 생성하고 있어서 콘크리트 클래스의 빈이 이중으로 생성되는 문제가 있었네요 👍

kang-hyungu reviewed (changes requested):
안녕하세요 이지 리뷰어 구구에요
재귀 대신 스택으로 빈 팩토리 구현한게 인상 깊네요
몇 가지 피드백 남겼는데 확인해보시고 리뷰 재요청주세요!

skygl committed:
refactor: 에러 테스트를 어노테이션 기반에서 패키지 기반으로 변경

- 테스트용 어노테이션을 만들었던 것을 패키지로 나누어서 리플렉션을 사용하도록 변경

skygl committed:
refactor: BeanScanner 의존성 수정

- 스캔할 어노테이션을 주입받는 것에서 BeanScanner 가 가지고 있도록 수정
- BeanFactory 가 BeanScanner 를 의존하지 않도록 수정

skygl committed:
refactor: AnnotationHandlerMapping 리팩토링

- initialize의 for문을 메소드로 추출

skygl committed:
refactor: CircularReferenceDetector 객체 생성 위치 변경

skygl committed:
fix: 인터페이스 Bean이 2개가 되는 에러 수정

- 인터페이스를 키로 갖는 빈 생성 방지 및 해당 경우 테스트 추가
- 빈을 만드는 로직 리팩토링

skygl committed:
refactor: UserAcceptanceTest에서 서버를 실행하도록 수정

- 기존 Disabled 어노테이션 제거

kang-hyungu reviewed (approved):
피드백 반영 잘 하셨어요 👍 
이번 미션 머지 처리할게요
수고하셨어요!
